vsintegration/src/FSharp.Editor/CodeLens/CodeLens.fs

```diff
@@ -0,0 +1,378 @@
+                [0..line.Length - 1] |> Seq.tryFind (fun i -> not (Char.IsWhiteSpace (line.Start.Add(i).GetChar())))
+                |> Option.defaultValue 0
+
+            let realStart = line.Start.Add(offset)
+            let span = SnapshotSpan(line.Snapshot, Span.FromBounds(int realStart, int line.End))
+            let geometry = view.TextViewLines.GetMarkerGeometry(span)
+            let! lens = codeLensResults.TryFind(text) // Check whether this line has code lens
+            if lens.Computed && codeLensUIElementCache.ContainsKey(text) then
+                // Use existing UI element which is proved to be safe to use
+                let textBox = codeLensUIElementCache.[text]
+                Canvas.SetLeft(textBox, geometry.Bounds.Left)
+                Canvas.SetTop(textBox, geometry.Bounds.Top - 15.)
+                return Some textBox
+            else if not lens.Computed then
+                lens.TaggedText |> Async.Ignore |> RoslynHelpers.StartAsyncSafe CancellationToken.None
+                let unusedElement = null
+                if codeLensUIElementCache.TryGetValue(text, &unusedElement) && not(isNull(unusedElement)) then
+                    return Some unusedElement
+                else
+                    return None
+            else
+                // No delay because it's already computed
+                let! taggedText, navigation = lens.TaggedText
+                
+                let textBox = new TextBlock(Width = view.ViewportWidth, Background = Brushes.Transparent, Opacity = 0.5, TextTrimming = TextTrimming.WordEllipsis)
+                DependencyObjectExtensions.SetDefaultTextProperties(textBox, formatMap.Value)
+                
+                for text in taggedText do
+                    let run = Documents.Run text.Text
+                    DependencyObjectExtensions.SetTextProperties (run, layoutTagToFormatting text.Tag)
+                    let inl =
+                        match text with
+                        | :? Layout.NavigableTaggedText as nav when navigation.IsTargetValid nav.Range ->                        
+                            let h = Documents.Hyperlink(run, ToolTip = nav.Range.FileName)
+                            h.Click.Add (fun _ -> navigation.NavigateTo nav.Range)
+                            h :> Documents.Inline
+                        | _ -> run :> _
+                    textBox.Inlines.Add inl
+
+                Canvas.SetLeft(textBox, geometry.Bounds.Left)
+                Canvas.SetTop(textBox, geometry.Bounds.Top - 15.)
+                textBox.Opacity <- 0.5
+                return Some textBox
+        } |> Async.map (fun ui ->
+               match ui with
+               | Some (Some ui) ->
+                   codeLensUIElementCache.TryAdd(text, ui) |> ignore
+                   Some ui
+               | _ -> 
+                   let unusedElement = null
+                   codeLensUIElementCache.TryRemove(text, &unusedElement) |> ignore
```

Comment:
I think this will also work
```fsharp
let (_,_) = codeLensUIElementCache.TryRemove text
```
(just another option if you don't want the `unusedElement` declaration)